augur/io/metadata.py

```diff
@@ -424,3 +426,47 @@ def write_records_to_tsv(records, output_file):
 
         for record in records:
             tsv_writer.writerow(record)
+
+
+class Metadata(TabularFile):
+    def __init__(self, file:str,
+            header:Optional[bool]=True,
+            names:Optional[List[str]]=None,
+            potential_id_columns:Optional[List[str]]=None):
+        super().__init__(file, header, names)
+
+        self.potential_id_columns = potential_id_columns or ("strain", "name")
+        """Metadata columns that may contain identifier information."""
+
+        self.id_column = _get_first_valid_column(self.file, self.potential_id_columns)
+        """Metadata ID column name."""
+
+
+def _get_first_valid_column(metadata_file:str, potential_column_names:List[str]):
+    """Returns the first column in `potential_column_names` that is present in the metadata.
+    Raises a `ValueError` when none of `potential_column_names` are found.
+    """
+    metadata_columns = _get_column_names(metadata_file)
+    for col in potential_column_names:
+        if col in metadata_columns:
+            return col
+    raise ValueError(f"Reading {metadata_file}: None of ({potential_column_names!r}) were found in the columns {tuple(metadata_columns)!r}.")
+
+
+def _get_column_names(file:str):
+    """Get column names of a tabular file."""
+    # Use the 'c' engine for faster parsing.
+    read_csv_kwargs = {
+        "sep": get_delimiter(file),
+        "engine": "c",
+        "skipinitialspace": True,
+        "dtype": 'string',
+    }
+
+    # Read the first row to detect columns.
+    row = pd.read_csv(
+        file,
+        nrows=1,
```

Comment:
I notice there is a lot of "peeking" at the first couple of rows in the metadata input for different purposes like this. In a single call we might:

 - sniff the delimiter
 - find the column names
 - find the column names and data types in pandas

Maybe it's best to keep these pieces logically distinct, although at first glance it feels like there could be a way to minimize the repeated inspection of the first few lines.

Reply:
This is a good point. With all the separation, there is some unnecessary duplication of `read_csv_kwargs`. This relates to your [comment](https://github.com/nextstrain/augur/pull/1094#discussion_r1030921512) on pandas-specific implementations and I'll address over there.

I believe keeping them separate is nice for readability (especially since delimiter detection is always necessary but not column name detection). The performance improvement from bundling into a single call is negligible since it is just "peeking".